sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java

```diff
@@ -110,16 +118,74 @@ public String apply(String input) {
+    final TupleTag<KV<String, String>> tableSchemasTag = new TupleTag<>();
+
+    List<PCollectionView<?>> resolveSideInputs = Lists.newArrayList();
+    if (dynamicDestinations.getSideInput() != null) {
+      resolveSideInputs.add(dynamicDestinations.getSideInput());
+    }
+    PCollectionTuple resolvedInputTuple = inputInGlobalWindow.apply(
+        "resolveInput",
+        ParDo.of(new DoFn<KV<DestinationT, TableRow>, KV<TableDestination, TableRow>>() {
+      DynamicDestinations<?, DestinationT> dynamicDestinationsCopy;
+
+      @StartBundle
+      public void startBundle(Context c) {
+        this.dynamicDestinationsCopy = SerializableUtils.clone(dynamicDestinations);
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        if (dynamicDestinationsCopy.getSideInput() != null) {
+          dynamicDestinationsCopy.setSideInputValue(
+              context.sideInput(dynamicDestinationsCopy.getSideInput()));
+        }
+        // Convert to a TableDestination and output.
+        TableDestination tableDestination =
+            dynamicDestinationsCopy.getTable(context.element().getKey());
+        TableReference tableReference = tableDestination.getTableReference();
+        if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
+          tableReference.setProjectId(
+              context.getPipelineOptions().as(BigQueryOptions.class).getProject());
+          tableDestination =
+              new TableDestination(
+                  tableReference, tableDestination.getTableDescription());
+        }
+        context.output(elementsOutputTag, KV.of(tableDestination, context.element().getValue()));
+
+        // Match the schema, and output the mapping.
+        TableSchema tableSchema = dynamicDestinationsCopy.getSchema(context.element().getKey());
+        if (tableSchema != null) {
+          context.output(tableSchemasTag, KV.of(tableDestination.getTableSpec(),
+              BigQueryHelpers.toJsonString(tableSchema)));
+        }
+      }
+    }).withSideInputs(resolveSideInputs)
+      .withOutputTags(elementsOutputTag, TupleTagList.of(tableSchemasTag)));
+
+    // We now have input in terms of TableDestinations.
+    PCollection<KV<TableDestination, TableRow>> resolvedInput =
+        resolvedInputTuple.get(elementsOutputTag)
+        .setCoder(KvCoder.of(TableDestinationCoder.of(), TableRowJsonCoder.of()));
+
+    // Produce the schema map to be used as a side input to following stages.
+    PCollectionView<Map<String, String>> schemaMapView =
+        resolvedInputTuple.get(tableSchemasTag)
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()))
+        .apply(Distinct.<KV<String, String>>create())
+        .apply(View.<String, String>asMap());
+
     // PCollection of filename, file byte size, and table destination.
     PCollection<WriteBundlesToFiles.Result> results =
-        inputInGlobalWindow
+        resolvedInput
```

Comment:
Wonder if it makes sense to propagate DestinationT all the way to WriteBundlesToFiles. I'm not sure.